### PR TITLE
feat(inventory grops): group name modal 

### PR DESCRIPTION
**Pull request by Fewwy**
Create group modal for inventory groups
image
image

To test the modal run "npx cypress open --component"
or you can add the CreateGroupModal to another component in the inventory. In this case, it will render but crash as soon as you make an API request for group name validation
OR you can open it in stage-beta
by opening the https://stage.foo.redhat.com:1337/beta/insights/inventory/groups and clicking on the "Create group" button

### CODE REVIEW

File: src/components/InventoryGroups/Modals/CreateGroupModal.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    // async validator has to throw error, not return it
+        throw 'Group name already exists';
+    } */
+};
+
+const validatorMapper = {
+    groupName: () => asyncGroupNameValidation
+};
+
+const createGroupSchema = {
```
Comment (mkholjuraev) [inline src/components/InventoryGroups/Modals/CreateGroupModal.js:28]: How about extracting modal assets into a new file?

File: src/components/InventoryGroups/Modals/CreateGroupModal.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+};
+
+const CreateGroupModal = ({
+    setIsModalOpen,
+    deviceIds,
+    reloadData
+}) => {
+    const dispatch = useDispatch();
+
+    const handleCreateGroup = (values) => {
```
Comment (mkholjuraev) [inline src/components/InventoryGroups/Modals/CreateGroupModal.js:56]: How about wrapping functions as this that are passed as a prop with `useCallback`? Maybe I use hooks heavily, but this helps to keep good performance.

**Review by mkholjuraev — COMMENTED** [review]
I am not familiar with the task, thus leaving only comments. 

Just as a reminder, I would like to note that these modals need to be loaded lazily wherever they are used. Even though the modals seem tiny, they import another library `data-driven-forms` which is big enough to affect performance.

File: src/components/InventoryGroups/Modals/CreateGroupModal.js (hunk 1 above)
Comment (Fewwy) [reply src/components/InventoryGroups/Modals/CreateGroupModal.js:30]: Good suggestion. will do that :)

File: src/components/InventoryGroups/Modals/CreateGroupModal.js (hunk 2 above)
Comment (Fewwy) [reply src/components/InventoryGroups/Modals/CreateGroupModal.js]: Added

File: src/components/InventoryGroups/Modals/CreateGroupModal.js (hunk 1 above)
Comment (Fewwy) [reply src/components/InventoryGroups/Modals/CreateGroupModal.js]: Moved the assets into the separate folder and file

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    it('Modal exist', () => {
+        cy.get(MODAL).should('exist');
+    });
+    it('Input exist', () => {
+        cy.get(TEXT_INPUT).should('exist');
+    });
+
+    it('Input is fillable and firing a validation request', () => {
+        cy.get(TEXT_INPUT).type('query');
+        cy.intercept('*', {
```
Comment (gkarat) [inline src/components/InventoryGroups/Modals/CreateGroupModal.cy.js:33]: The URL you wanna catch must be more precious

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        }).as('validate');
+        cy.wait('@validate').then((xhr) => {
+            expect(xhr.request.url).to.contain('groups/query');}
+        );
+    });
+
+    it('Input is fillable and firing a create group', () => {
+        cy.get(TEXT_INPUT).type('group_name');
+        cy.get(`button[type="submit"]`).click();
+        cy.intercept('*', {
```
Comment (gkarat) [inline src/components/InventoryGroups/Modals/CreateGroupModal.cy.js:44]: The same here

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js (hunk 3 above)
Comment (gkarat) [reply src/components/InventoryGroups/Modals/CreateGroupModal.cy.js:33]: And the status code is not 201 here

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    });
+
+    it('Input is fillable and firing a validation request', () => {
+        cy.get(TEXT_INPUT).type('query');
+        cy.intercept('*', {
+            statusCode: 201
+        }).as('validate');
+        cy.wait('@validate').then((xhr) => {
+            expect(xhr.request.url).to.contain('groups/query');}
+        );
```
Comment (gkarat) [inline src/components/InventoryGroups/Modals/CreateGroupModal.cy.js:38]:
```suggestion
        cy.wait('@validate');
```

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    it('Input is fillable and firing a create group', () => {
+        cy.get(TEXT_INPUT).type('group_name');
+        cy.get(`button[type="submit"]`).click();
+        cy.intercept('*', {
+            statusCode: 201
+        }).as('create_group');
+        cy.wait('@create_group').then((xhr) => {
+            expect(xhr.request.url).to.contain('groups');}
+        );
+    });
```
Comment (gkarat) [inline src/components/InventoryGroups/Modals/CreateGroupModal.cy.js:50]:
```suggestion
        cy.wait('@create_group');
    });
```

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+                </Provider>
+            </MemoryRouter>
+        );
+    });
+    it('Modal exist', () => {
+        cy.get(MODAL).should('exist');
+    });
+    it('Input exist', () => {
+        cy.get(TEXT_INPUT).should('exist');
+    });
```
Comment (gkarat) [inline src/components/InventoryGroups/Modals/CreateGroupModal.cy.js:29]: This must go to unit tests (react-testing-library)

**Review by gkarat — CHANGES_REQUESTED** [review]

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js (hunk 5 above)
Comment (Fewwy) [reply src/components/InventoryGroups/Modals/CreateGroupModal.cy.js:45]: Updated

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js (hunk 4 above)
Comment (Fewwy) [reply src/components/InventoryGroups/Modals/CreateGroupModal.cy.js]: Updatedd=

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js (hunk 3 above)
Comment (Fewwy) [reply src/components/InventoryGroups/Modals/CreateGroupModal.cy.js]: Done

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js (hunk 6 above)
Comment (Fewwy) [reply src/components/InventoryGroups/Modals/CreateGroupModal.cy.js]: Done

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js (hunk 7 above)
Comment (Fewwy) [reply src/components/InventoryGroups/Modals/CreateGroupModal.cy.js]: Done

File: src/components/InventoryGroups/utils/api.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import { instance } from '@redhat-cloud-services/frontend-components-utilities/interceptors/interceptors';
+import { INVENTORY_API_BASE } from '../../../api';
+
+//had to replace the instance.post with axios.post otherwise I had an error because chrome is missing
```
Comment (gkarat) [inline src/components/InventoryGroups/utils/api.js:4]: The comment doesn't seem to be appropriate here (you use instance.post)

File: src/components/InventoryGroups/utils/api.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+export const addSystemsToGroup = (groupId, systems) => {
+    return instance.post(`${INVENTORY_API_BASE}/groups/${groupId}`, {
+        ID: groupId,
+        Devices: systems
+    });
+};
+
+export const validateGroupName = (name) => {
+    return instance.get(`${INVENTORY_API_BASE}/groups/${name}`);
```
Comment (gkarat) [inline src/components/InventoryGroups/utils/api.js:20]: There is no such endpoint in the API spec. You can only perform GET /groups/ID with ID, not name.

**Review by gkarat — CHANGES_REQUESTED** [review]

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        cy.window().then(window => window.insights = {
+            chrome: {
+                isProd: false,
+                auth: {
+                    getUser: () => {
+                        return Promise.resolve({});
+                    }
+                }
+            }
+        });
```
[the remark this answers is not in the document]
Comment (Fewwy) [reply src/components/InventoryGroups/Modals/CreateGroupModal.cy.js:22]: Good suggestion, added it

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        cy.intercept('GET', '**/api/inventory/v1/groups/query', {
```
[the remark this answers is not in the document]
Comment (Fewwy) [reply src/components/InventoryGroups/Modals/CreateGroupModal.cy.js:23]: You are correct I need to rewrite the validation to make it work with a returned list of items

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Fewwy) [reply src/components/InventoryGroups/Modals/CreateGroupModal.cy.js:22]: Great suggestion, I implemented it

File: src/components/InventoryGroups/Modals/CreateGroupModal.cy.js [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+            statusCode: 200
+        }).as('validate');
+        cy.intercept('POST', '**/api/inventory/v1/groups', {
+            statusCode: 504
+        }).as('create_group');
+        cy.intercept('GET', '**/api/inventory/v1/groups/group_name', {
```
[the remark this answers is not in the document]
Comment (Fewwy) [reply src/components/InventoryGroups/Modals/CreateGroupModal.cy.js:29]: removed that and updated the API fetch functions to match the specs

File: src/components/InventoryGroups/Modals/CreateGroupModal.js [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+import { createGroupSchema } from './ModalSchemas/schemes';
+import Modal from './Modal';
+import apiWithToast from '../utils/apiWithToast';
+import {
+    createGroup,
+    addSystemsToGroup,
+    validateGroupName
+} from '../utils/api';
+import { useDispatch } from 'react-redux';
+import { debounce } from 'lodash';
```
[the remark this answers is not in the document]
Comment (Fewwy) [reply src/components/InventoryGroups/Modals/CreateGroupModal.js:12]: great suggestion

File: src/components/InventoryGroups/utils/api.js (hunk 8 above)
Comment (Fewwy) [reply src/components/InventoryGroups/utils/api.js:4]: Removed it

File: src/components/InventoryGroups/Modals/Modal.js [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+                            submit: { variant }
+                        }}
+                    />
+                )}
+                initialValues={initialValues}
+                componentMapper={
+                    additionalMappers
+                        ? { ...additionalMappers, ...componentMapper }
+                        : componentMapper
+                }
```
[the remark this answers is not in the document]
Comment (Fewwy) [reply src/components/InventoryGroups/Modals/Modal.js:50]: Form won't render without the component mapper, I will remove additional mappers but keep component mapper

File: src/components/InventoryGroups/utils/api.js [hunk 15]
```diff
@@ -0,0 +1,21 @@
+import { instance } from '@redhat-cloud-services/frontend-components-utilities/interceptors/interceptors';
+import { INVENTORY_API_BASE } from '../../../api';
+
+//had to replace the instance.post with axios.post otherwise I had an error because chrome is missing
+export const createGroup = (payload) => {
+    return instance.post(`${INVENTORY_API_BASE}/groups`, {
+        Name: payload.name,
```
[the remark this answers is not in the document]
Comment (Fewwy) [reply src/components/InventoryGroups/utils/api.js:7]: updated the api create group funciton

File: src/components/InventoryGroups/utils/api.js [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+        Type: 'static'
+    });
+};
+
+export const addSystemsToGroup = (groupId, systems) => {
```
[the remark this answers is not in the document]
Comment (Fewwy) [reply src/components/InventoryGroups/utils/api.js:12]: Removed it

File: src/components/InventoryGroups/utils/api.js (hunk 9 above)
Comment (Fewwy) [reply src/components/InventoryGroups/utils/api.js:20]: Updated the validate group name function

**Review by gkarat — COMMENTED** [review]
The modal still doesn't check for existing groups correctly :-( Attached the video:

[Screencast from 02-15-2023 04:06:55 PM.webm](https://user-images.githubusercontent.com/31385370/219066608-d8511b73-e19a-4244-ba5b-eb99930f9174.webm)

File: src/components/InventoryGroups/utils/api.js [hunk 17]
```diff
@@ -1,6 +1,19 @@
...
+    return instance.post(`${INVENTORY_API_BASE}/groups`, {
+        name: payload.name,
+        // eslint-disable-next-line camelcase
+        host_ids: []
+    });
+};
+
+export const validateGroupName = (name) => {
+    return instance.get(`${INVENTORY_API_BASE}/groups`)
+    .then((resp) => resp[0]?.results.some((group) => group.name === name));
```
Comment (gkarat) [inline src/components/InventoryGroups/utils/api.js:14]:
```suggestion
    .then((resp) => resp?.results.some((group) => group.name === name));
```
```suggestion
    .then((resp) => resp[0]?.results.some((group) => group.name === name));
```

**Review by gkarat — CHANGES_REQUESTED** [review]

**Review by gkarat — APPROVED** [review]
this looks great! run locally and all the validators work. also the correct requests are sent to validate/create a group.

let's address the validator function mentioned above. other than that, LGTM.

**Pull request merged**